core/chain_makers.go

```diff
@@ -340,7 +366,102 @@ func (cr *fakeChainReader) GetGasUsedInChain(block *types.Block, length int) int
+	// the first config must be the Prime
+	if primeConfig != params.MainnetPrimeChainConfig {
+		print("must have MainnetPrimeChainConfig as first config")
+	}
+	// check second config for Region
+	if !(regionConfig.ChainID.Cmp(params.MainnetRegionChainConfigs[0].ChainID) == 0 || regionConfig.ChainID.Cmp(params.MainnetRegionChainConfigs[1].ChainID) == 0 || regionConfig.ChainID.Cmp(params.MainnetRegionChainConfigs[2].ChainID) == 0) {
+		print("must have a MainnetRegionChainConfig as second config")
+	}
+	// check third config for Zone in Region
+	if !(zoneConfig.ChainID.Cmp(params.MainnetZoneChainConfigs[regionConfig.Location[0]-1][0].ChainID) == 0 || zoneConfig.ChainID.Cmp(params.MainnetZoneChainConfigs[regionConfig.Location[0]-1][1].ChainID) == 0 || zoneConfig.ChainID.Cmp(params.MainnetZoneChainConfigs[regionConfig.Location[0]-1][2].ChainID) == 0) {
+		print("must have a MainnetZoneConfig as third config")
+	}
+
+	n := len(orders)
+	blocks, receipts := make(types.Blocks, n), make([]types.Receipts, n)
+
+	// associate chainreaders and configs for correct block context placement
+	chainreaders, configs := []fakeChainReader{}, []params.ChainConfig{}
+	for _, context := range orders {
+		switch context {
+		case 0:
+			chainreaders = append(chainreaders, fakeChainReader{config: primeConfig})
+			configs = append(configs, *primeConfig)
+		case 1:
+			chainreaders = append(chainreaders, fakeChainReader{config: &regionConfig})
+			configs = append(configs, regionConfig)
+		case 2:
+			chainreaders = append(chainreaders, fakeChainReader{config: &zoneConfig})
+			configs = append(configs, zoneConfig)
+		default:
+			print("contexts must be 0, 1, or 2")
+		}
+	}
+
+	genblock := func(i int, parent *types.Block, context int, statedb *state.StateDB, chainreaders []fakeChainReader, configs []params.ChainConfig) (*types.Block, types.Receipts) {
+		// select appropriate fakeChainReader for block creation
+		chainreader := chainreaders[i]
+		config := configs[i]
+
+		b := &BlockGen{i: i, chain: blocks, parent: parent, statedb: statedb, config: &config, engine: engine}
+		b.header = makeHeader(&config, &chainreader, parent, context, statedb, b.engine)
+
+		// Execute any user modifications to the block
+		if gen != nil {
+			gen(i, b)
+		}
+
+		if b.engine != nil {
+			// Finalize and seal the block
+			block, _ := b.engine.FinalizeAndAssemble(&chainreader, b.header, statedb, b.txs, b.uncles, b.receipts)
+
+			// Write state changes to db
+			root, err := statedb.Commit(config.IsEIP158(b.header.Number[types.QuaiNetworkContext]))
+			if err != nil {
+				panic(fmt.Sprintf("state write error: %v", err))
+			}
+			if err := statedb.Database().TrieDB().Commit(root, false, nil); err != nil {
+				panic(fmt.Sprintf("trie write error: %v", err))
+			}
+			return block, b.receipts
+		}
+		return nil, nil
+	}
+
+	for i := 0; i < n; i++ {
+		statedb, err := state.New(parent.Root(), state.NewDatabase(db), nil)
```

Comment:
Does this line create a new database for each block that is generated? Or a new state variable referencing the same database? Either way that doesn't sound right. Wouldn't you just have one chain state with one db, and you put `n` blocks into that one chain?

Reply:
That is a good question and I am vague on some of the finer details how state relates to the database object/s. I will make a point to look in to it in case there is a more efficient way to update/handle state.